File: archive/swift/Sources/GedFixApp/Models/GedcomParser.swift
import Foundation

/// Result of parsing a GEDCOM file
struct GedcomParseResult: Sendable {
    var persons: [GedcomPerson]
    var families: [GedcomFamily]
    var childLinks: [GedcomChildLink]
    var events: [GedcomEvent]
    var places: [GedcomPlace]
    var sources: [GedcomSource]
    var lineCount: Int
}

/// A single parsed GEDCOM line
private struct GedLine {
    let level: Int
    let xref: String?  // e.g. @I123@
    let tag: String
    let value: String
}

/// Line-oriented GEDCOM 5.5.1 parser
/// Ported from gedfix/checker.py (Python)
enum GedcomParser {

    // MARK: - Public API

    static func parse(fileURL: URL) throws -> GedcomParseResult {
        let data = try Data(contentsOf: fileURL)
        // Handle BOM
        let text: String
        if data.starts(with: [0xEF, 0xBB, 0xBF]) {
            text = String(data: data.dropFirst(3), encoding: .utf8) ?? ""
        } else {
            text = String(data: data, encoding: .utf8)
                ?? String(data: data, encoding: .isoLatin1)
                ?? ""
        }

        let rawLines = text.components(separatedBy: .newlines)
        let parsedLines = rawLines.compactMap { parseLine($0) }
        let records = groupRecords(parsedLines)

        var persons: [GedcomPerson] = []
        var families: [GedcomFamily] = []
        var childLinks: [GedcomChildLink] = []
        var events: [GedcomEvent] = []
        var placesDict: [String: GedcomPlace] = [:]
        var sources: [GedcomSource] = []

        for record in records {
            guard let header = record.first else { continue }

            if header.tag == "INDI", let xref = header.xref {
                let (person, personEvents) = parseINDI(xref: xref, lines: record)
                persons.append(person)
                for evt in personEvents {
                    events.append(evt)
                    if !evt.place.isEmpty {
                        trackPlace(evt.place, in: &placesDict)
                    }
                }
            } else if header.tag == "FAM", let xref = header.xref {
                let (family, links, famEvents) = parseFAM(xref: xref, lines: record)
                families.append(family)
                childLinks.append(contentsOf: links)
                for evt in famEvents {
                    events.append(evt)
                    if !evt.place.isEmpty {
                        trackPlace(evt.place, in: &placesDict)
                    }
                }
            } else if header.tag == "SOUR", let xref = header.xref {
                let source = parseSOUR(xref: xref, lines: record)
                sources.append(source)
            }
        }

        // Detect living persons
        let currentYear = Calendar.current.component(.year, from: Date())
        for i in persons.indices {
            let xref = persons[i].xref
            let personEvents = events.filter { $0.ownerXref == xref }
            let hasDeath = personEvents.contains { $0.eventType == "DEAT" || $0.eventType == "BURI" }

            if hasDeath {
                persons[i].isLiving = false
            } else {
                let birthYear = personEvents
                    .filter { $0.eventType == "BIRT" }
                    .compactMap { extractYear($0.dateValue) }
                    .first

                if let by = birthYear, (currentYear - by) > 110 {
                    persons[i].isLiving = false
                } else {
                    persons[i].isLiving = true
                }
            }
        }

        return GedcomParseResult(
            persons: persons,
            families: families,
            childLinks: childLinks,
            events: events,
            places: Array(placesDict.values),
            sources: sources,
            lineCount: rawLines.count
        )
    }

    // MARK: - Line Parsing

    private static func parseLine(_ raw: String) -> GedLine? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        let parts = trimmed.split(separator: " ", maxSplits: 2, omittingEmptySubsequences: true)
        guard parts.count >= 2, let level = Int(parts[0]) else { return nil }

        let second = String(parts[1])

        // Check if second token is an xref (@X@)
        if second.hasPrefix("@") && second.hasSuffix("@") && parts.count >= 3 {
            let xref = second
            let tag = String(parts[2].split(separator: " ", maxSplits: 1).first ?? "")
            let value = parts[2].split(separator: " ", maxSplits: 1).dropFirst().first.map(String.init) ?? ""
            return GedLine(level: level, xref: xref, tag: tag, value: value)
        }

        let tag = second
        let value = parts.count >= 3 ? String(parts[2]) : ""
        return GedLine(level: level, xref: nil, tag: tag, value: value)
    }

    // MARK: - Record Grouping

    private static func groupRecords(_ lines: [GedLine]) -> [[GedLine]] {
        var records: [[GedLine]] = []
        var current: [GedLine] = []

        for line in lines {
            if line.level == 0 {
                if !current.isEmpty {
                    records.append(current)
                }
                current = [line]
            } else {
                current.append(line)
            }
        }
        if !current.isEmpty {
            records.append(current)
        }
        return records
    }

    // MARK: - INDI Parsing

    private static func parseINDI(xref: String, lines: [GedLine]) -> (GedcomPerson, [GedcomEvent]) {
        var givenName = ""
        var surname = ""
        var suffix = ""
        var sex = "U"
        var events: [GedcomEvent] = []

        var currentTag: String?
        var currentDate = ""
        var currentPlace = ""

        for line in lines.dropFirst() {
            if line.level == 1 {
                // Flush previous event
                if let tag = currentTag, isEventTag(tag) {
                    events.append(makeEvent(owner: xref, ownerType: "INDI", type: tag, date: currentDate, place: currentPlace))
                }
                currentTag = line.tag
                currentDate = ""
                currentPlace = ""

                switch line.tag {
                case "NAME":
                    let parsed = parseName(line.value)
                    if givenName.isEmpty {
                        givenName = parsed.given
                        surname = parsed.surname
                        suffix = parsed.suffix
                    }
                case "SEX":
                    sex = line.value.trimmingCharacters(in: .whitespaces)
                default:
                    break
                }
            } else if line.level == 2 {
                switch line.tag {
                case "DATE":
                    currentDate = line.value
                case "PLAC":
                    currentPlace = line.value
                case "GIVN":
                    if givenName.isEmpty { givenName = line.value }
                case "SURN":
                    if surname.isEmpty { surname = line.value }
                case "NSFX":
                    if suffix.isEmpty { suffix = line.value }
                default:
                    break
                }
            }
        }

        // Flush last event
        if let tag = currentTag, isEventTag(tag) {
            events.append(makeEvent(owner: xref, ownerType: "INDI", type: tag, date: currentDate, place: currentPlace))
        }

        let person = GedcomPerson(
            id: UUID().uuidString,
            xref: xref,
            givenName: givenName.trimmingCharacters(in: .whitespaces),
            surname: surname.trimmingCharacters(in: .whitespaces),
            suffix: suffix.trimmingCharacters(in: .whitespaces),
            sex: sex,
            isLiving: false  // set later
        )
        return (person, events)
    }

    // MARK: - FAM Parsing

    private static func parseFAM(xref: String, lines: [GedLine]) -> (GedcomFamily, [GedcomChildLink], [GedcomEvent]) {
        var husbXref = ""
        var wifeXref = ""
        var children: [String] = []
        var events: [GedcomEvent] = []

        var currentTag: String?
        var currentDate = ""
        var currentPlace = ""

        for line in lines.dropFirst() {
            if line.level == 1 {
                // Flush event
                if let tag = currentTag, isEventTag(tag) {
                    events.append(makeEvent(owner: xref, ownerType: "FAM", type: tag, date: currentDate, place: currentPlace))
                }
                currentTag = line.tag
                currentDate = ""
                currentPlace = ""

                switch line.tag {
                case "HUSB":
                    husbXref = line.value
                case "WIFE":
                    wifeXref = line.value
                case "CHIL":
                    children.append(line.value)
                default:
                    break
                }
            } else if line.level == 2 {
                if line.tag == "DATE" { currentDate = line.value }
                if line.tag == "PLAC" { currentPlace = line.value }
            }
        }

        if let tag = currentTag, isEventTag(tag) {
            events.append(makeEvent(owner: xref, ownerType: "FAM", type: tag, date: currentDate, place: currentPlace))
        }

        let family = GedcomFamily(
            id: UUID().uuidString,
            xref: xref,
            partner1Xref: husbXref,
            partner2Xref: wifeXref
        )

        let links = children.enumerated().map { (i, childXref) in
            GedcomChildLink(familyXref: xref, childXref: childXref, childOrder: i)
        }

        return (family, links, events)
    }

    // MARK: - SOUR Parsing

    private static func parseSOUR(xref: String, lines: [GedLine]) -> GedcomSource {
        var title = ""
        var author = ""
        var publisher = ""
        var repository = ""

        for line in lines.dropFirst() {
            if line.level == 1 {
                switch line.tag {
                case "TITL": title = line.value
                case "AUTH": author = line.value
                case "PUBL": publisher = line.value
                case "REPO": repository = line.value
                default: break
                }
            }
        }

        return GedcomSource(
            id: UUID().uuidString,
            xref: xref,
            title: title,
            author: author,
            publisher: publisher,
            repository: repository
        )
    }

    // MARK: - Helpers

    private static func parseName(_ raw: String) -> (given: String, surname: String, suffix: String) {
        // GEDCOM name format: "Given /Surname/ Suffix"
        let parts = raw.split(separator: "/", omittingEmptySubsequences: false)
        if parts.count >= 2 {
            let given = String(parts[0]).trimmingCharacters(in: .whitespaces)
            let surname = String(parts[1]).trimmingCharacters(in: .whitespaces)
            let suffix = parts.count >= 3 ? String(parts[2]).trimmingCharacters(in: .whitespaces) : ""
            return (given, surname, suffix)
        }
        return (raw.trimmingCharacters(in: .whitespaces), "", "")
    }

    private static let eventTags: Set<String> = [
        "BIRT", "DEAT", "MARR", "BURI", "CHR", "BAPM", "RESI", "CENS",
        "IMMI", "EMIG", "NATU", "GRAD", "RETI", "PROB", "WILL", "DIV",
        "EVEN", "CREM", "ADOP", "CONF", "FCOM", "ORDN", "ANUL"
    ]

    private static func isEventTag(_ tag: String) -> Bool {
        eventTags.contains(tag)
    }

    private static func makeEvent(owner: String, ownerType: String, type: String, date: String, place: String) -> GedcomEvent {
        GedcomEvent(
            id: UUID().uuidString,
            ownerXref: owner,
            ownerType: ownerType,
            eventType: type,
            dateValue: date.trimmingCharacters(in: .whitespaces),
            place: place.trimmingCharacters(in: .whitespaces),
            description: ""
        )
    }

    private static func trackPlace(_ name: String, in dict: inout [String: GedcomPlace]) {
        if var existing = dict[name] {
            existing.eventCount += 1
            dict[name] = existing
        } else {
            dict[name] = GedcomPlace(
                id: UUID().uuidString,
                name: name,
                normalized: name,
                latitude: nil,
                longitude: nil,
                eventCount: 1
            )
        }
    }

    private static func extractYear(_ dateString: String) -> Int? {
        let pattern = /(\d{4})/
        guard let match = dateString.firstMatch(of: pattern) else { return nil }
        return Int(match.1)
    }
}

File: archive/kmp/composeApp/src/commonMain/kotlin/com/gedfix/models/GedcomParser.kt
package com.gedfix.models

import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * A single parsed GEDCOM line.
 */
private data class GedLine(
    val level: Int,
    val xref: String?, // e.g. @I123@
    val tag: String,
    val value: String
)

/**
 * Line-oriented GEDCOM 5.5.1 parser.
 * Ported from Swift GedcomParser.swift / gedfix/checker.py
 */
@OptIn(ExperimentalUuidApi::class)
object GedcomParser {

    // MARK: - Public API

    fun parse(text: String, currentYear: Int = 2026): GedcomParseResult {
        // Handle BOM
        val cleaned = if (text.startsWith("\uFEFF")) text.removePrefix("\uFEFF") else text

        val rawLines = cleaned.lines()
        val parsedLines = rawLines.mapNotNull { parseLine(it) }
        val records = groupRecords(parsedLines)

        val persons = mutableListOf<GedcomPerson>()
        val families = mutableListOf<GedcomFamily>()
        val childLinks = mutableListOf<GedcomChildLink>()
        val events = mutableListOf<GedcomEvent>()
        val placesDict = mutableMapOf<String, GedcomPlace>()
        val sources = mutableListOf<GedcomSource>()
        val media = mutableListOf<GedcomMedia>()

        for (record in records) {
            val header = record.firstOrNull() ?: continue

            when {
                header.tag == "INDI" && header.xref != null -> {
                    val (person, personEvents, personMedia) = parseINDI(xref = header.xref, lines = record)
                    persons.add(person)
                    for (evt in personEvents) {
                        events.add(evt)
                        if (evt.place.isNotEmpty()) {
                            trackPlace(evt.place, placesDict)
                        }
                    }
                    media.addAll(personMedia)
                }
                header.tag == "FAM" && header.xref != null -> {
                    val (family, links, famEvents) = parseFAM(xref = header.xref, lines = record)
                    families.add(family)
                    childLinks.addAll(links)
                    for (evt in famEvents) {
                        events.add(evt)
                        if (evt.place.isNotEmpty()) {
                            trackPlace(evt.place, placesDict)
                        }
                    }
                }
                header.tag == "SOUR" && header.xref != null -> {
                    val source = parseSOUR(xref = header.xref, lines = record)
                    sources.add(source)
                }
                header.tag == "OBJE" && header.xref != null -> {
                    val mediaItem = parseOBJE(xref = header.xref, ownerXref = "", lines = record)
                    media.add(mediaItem)
                }
            }
        }

        // Detect living persons
        for (i in persons.indices) {
            val xref = persons[i].xref
            val personEvents = events.filter { it.ownerXref == xref }
            val hasDeath = personEvents.any { it.eventType == "DEAT" || it.eventType == "BURI" }

            if (hasDeath) {
                persons[i] = persons[i].copy(isLiving = false)
            } else {
                val birthYear = personEvents
                    .filter { it.eventType == "BIRT" }
                    .mapNotNull { extractYear(it.dateValue) }
                    .firstOrNull()

                persons[i] = if (birthYear != null && (currentYear - birthYear) > 110) {
                    persons[i].copy(isLiving = false)
                } else {
                    persons[i].copy(isLiving = true)
                }
            }
        }

        return GedcomParseResult(
            persons = persons.toList(),
            families = families.toList(),
            childLinks = childLinks.toList(),
            events = events.toList(),
            places = placesDict.values.toList(),
            sources = sources.toList(),
            media = media.toList(),
            lineCount = rawLines.size
        )
    }

    // MARK: - Line Parsing

    private fun parseLine(raw: String): GedLine? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null

        val parts = trimmed.split(" ", limit = 3)
        if (parts.size < 2) return null

        val level = parts[0].toIntOrNull() ?: return null
        val second = parts[1]

        // Check if second token is an xref (@X@)
        if (second.startsWith("@") && second.endsWith("@") && parts.size >= 3) {
            val xref = second
            val remaining = parts[2]
            val tagParts = remaining.split(" ", limit = 2)
            val tag = tagParts[0]
            val value = if (tagParts.size >= 2) tagParts[1] else ""
            return GedLine(level = level, xref = xref, tag = tag, value = value)
        }

        val tag = second
        val value = if (parts.size >= 3) parts[2] else ""
        return GedLine(level = level, xref = null, tag = tag, value = value)
    }

    // MARK: - Record Grouping

    private fun groupRecords(lines: List<GedLine>): List<List<GedLine>> {
        val records = mutableListOf<List<GedLine>>()
        var current = mutableListOf<GedLine>()

        for (line in lines) {
            if (line.level == 0) {
                if (current.isNotEmpty()) {
                    records.add(current.toList())
                }
                current = mutableListOf(line)
            } else {
                current.add(line)
            }
        }
        if (current.isNotEmpty()) {
            records.add(current.toList())
        }
        return records
    }

    // MARK: - INDI Parsing

    private fun parseINDI(xref: String, lines: List<GedLine>): Triple<GedcomPerson, List<GedcomEvent>, List<GedcomMedia>> {
        var givenName = ""
        var surname = ""
        var suffix = ""
        var sex = "U"
        val events = mutableListOf<GedcomEvent>()
        val mediaItems = mutableListOf<GedcomMedia>()
        var sourCount = 0
        var obiCount = 0

        var currentTag: String? = null
        var currentDate = ""
        var currentPlace = ""

        // Track inline OBJE sub-tags
        var inObje = false
        var objeFile = ""
        var objeForm = ""
        var objeTitl = ""
        var objeNote = ""

        for (line in lines.drop(1)) {
            if (line.level == 1) {
                // Flush previous event
                currentTag?.let { tag ->
                    if (isEventTag(tag)) {
                        events.add(makeEvent(owner = xref, ownerType = "INDI", type = tag, date = currentDate, place = currentPlace))
                    }
                }
                // Flush previous inline OBJE
                if (inObje && objeFile.isNotEmpty()) {
                    mediaItems.add(GedcomMedia(
                        id = Uuid.random().toString(),
                        xref = "",
                        ownerXref = xref,
                        filePath = objeFile,
                        format = objeForm,
                        title = objeTitl,
                        description = objeNote
                    ))
                }
                inObje = false
                objeFile = ""
                objeForm = ""
                objeTitl = ""
                objeNote = ""

                currentTag = line.tag
                currentDate = ""
                currentPlace = ""

                when (line.tag) {
                    "NAME" -> {
                        val parsed = parseName(line.value)
                        if (givenName.isEmpty()) {
                            givenName = parsed.first
                            surname = parsed.second
                            suffix = parsed.third
                        }
                    }
                    "SEX" -> {
                        sex = line.value.trim()
                    }
                    "SOUR" -> sourCount++
                    "OBJE" -> {
                        obiCount++
                        inObje = true
                    }
                }
            } else if (line.level == 2) {
                if (inObje) {
                    when (line.tag) {
                        "FILE" -> objeFile = line.value
                        "FORM" -> objeForm = line.value
                        "TITL" -> objeTitl = line.value
                        "NOTE" -> objeNote = line.value
                    }
                } else {
                    when (line.tag) {
                        "DATE" -> currentDate = line.value
                        "PLAC" -> currentPlace = line.value
                        "GIVN" -> if (givenName.isEmpty()) givenName = line.value
                        "SURN" -> if (surname.isEmpty()) surname = line.value
                        "NSFX" -> if (suffix.isEmpty()) suffix = line.value
                        "SOUR" -> sourCount++
                        "OBJE" -> obiCount++
                    }
                }
            }
        }

        // Flush last event
        currentTag?.let { tag ->
            if (isEventTag(tag)) {
                events.add(makeEvent(owner = xref, ownerType = "INDI", type = tag, date = currentDate, place = currentPlace))
            }
        }
        // Flush last inline OBJE
        if (inObje && objeFile.isNotEmpty()) {
            mediaItems.add(GedcomMedia(
                id = Uuid.random().toString(),
                xref = "",
                ownerXref = xref,
                filePath = objeFile,
                format = objeForm,
                title = objeTitl,
                description = objeNote
            ))
        }

        val person = GedcomPerson(
            id = Uuid.random().toString(),
            xref = xref,
            givenName = givenName.trim(),
            surname = surname.trim(),
            suffix = suffix.trim(),
            sex = sex,
            isLiving = false, // set later
            sourceCount = sourCount,
            mediaCount = obiCount,
            isValidated = sourCount > 0
        )
        return Triple(person, events, mediaItems)
    }

    // MARK: - FAM Parsing

    private fun parseFAM(xref: String, lines: List<GedLine>): Triple<GedcomFamily, List<GedcomChildLink>, List<GedcomEvent>> {
        var husbXref = ""
        var wifeXref = ""
        val children = mutableListOf<String>()
        val events = mutableListOf<GedcomEvent>()

        var currentTag: String? = null
        var currentDate = ""
        var currentPlace = ""

        for (line in lines.drop(1)) {
            if (line.level == 1) {
                // Flush event
                currentTag?.let { tag ->
                    if (isEventTag(tag)) {
                        events.add(makeEvent(owner = xref, ownerType = "FAM", type = tag, date = currentDate, place = currentPlace))
                    }
                }
                currentTag = line.tag
                currentDate = ""
                currentPlace = ""

                when (line.tag) {
                    "HUSB" -> husbXref = line.value
                    "WIFE" -> wifeXref = line.value
                    "CHIL" -> children.add(line.value)
                }
            } else if (line.level == 2) {
                if (line.tag == "DATE") currentDate = line.value
                if (line.tag == "PLAC") currentPlace = line.value
            }
        }

        currentTag?.let { tag ->
            if (isEventTag(tag)) {
                events.add(makeEvent(owner = xref, ownerType = "FAM", type = tag, date = currentDate, place = currentPlace))
            }
        }

        val family = GedcomFamily(
            id = Uuid.random().toString(),
            xref = xref,
            partner1Xref = husbXref,
            partner2Xref = wifeXref
        )

        val links = children.mapIndexed { i, childXref ->
            GedcomChildLink(familyXref = xref, childXref = childXref, childOrder = i)
        }

        return Triple(family, links, events)
    }

    // MARK: - SOUR Parsing

    private fun parseSOUR(xref: String, lines: List<GedLine>): GedcomSource {
        var title = ""
        var author = ""
        var publisher = ""
        var repository = ""

        for (line in lines.drop(1)) {
            if (line.level == 1) {
                when (line.tag) {
                    "TITL" -> title = line.value
                    "AUTH" -> author = line.value
                    "PUBL" -> publisher = line.value
                    "REPO" -> repository = line.value
                }
            }
        }

        return GedcomSource(
            id = Uuid.random().toString(),
            xref = xref,
            title = title,
            author = author,
            publisher = publisher,
            repository = repository
        )
    }

    // MARK: - OBJE Parsing

    private fun parseOBJE(xref: String, ownerXref: String, lines: List<GedLine>): GedcomMedia {
        var filePath = ""
        var format = ""
        var title = ""
        var description = ""

        for (line in lines.drop(1)) {
            if (line.level == 1) {
                when (line.tag) {
                    "FILE" -> filePath = line.value
                    "FORM" -> format = line.value
                    "TITL" -> title = line.value
                    "NOTE" -> description = line.value
                }
            } else if (line.level == 2) {
                when (line.tag) {
                    "FORM" -> if (format.isEmpty()) format = line.value
                    "TITL" -> if (title.isEmpty()) title = line.value
                }
            }
        }

        return GedcomMedia(
            id = Uuid.random().toString(),
            xref = xref,
            ownerXref = ownerXref,
            filePath = filePath,
            format = format,
            title = title,
            description = description
        )
    }

    // MARK: - Helpers

    private fun parseName(raw: String): Triple<String, String, String> {
        // GEDCOM name format: "Given /Surname/ Suffix"
        val parts = raw.split("/")
        if (parts.size >= 2) {
            val given = parts[0].trim()
            val surname = parts[1].trim()
            val suffix = if (parts.size >= 3) parts[2].trim() else ""
            return Triple(given, surname, suffix)
        }
        return Triple(raw.trim(), "", "")
    }

    private val eventTags = setOf(
        "BIRT", "DEAT", "MARR", "BURI", "CHR", "BAPM", "RESI", "CENS",
        "IMMI", "EMIG", "NATU", "GRAD", "RETI", "PROB", "WILL", "DIV",
        "EVEN", "CREM", "ADOP", "CONF", "FCOM", "ORDN", "ANUL"
    )

    private fun isEventTag(tag: String): Boolean = tag in eventTags

    private fun makeEvent(owner: String, ownerType: String, type: String, date: String, place: String): GedcomEvent {
        return GedcomEvent(
            id = Uuid.random().toString(),
            ownerXref = owner,
            ownerType = ownerType,
            eventType = type,
            dateValue = date.trim(),
            place = place.trim(),
            description = ""
        )
    }

    private fun trackPlace(name: String, dict: MutableMap<String, GedcomPlace>) {
        val existing = dict[name]
        if (existing != null) {
            dict[name] = existing.copy(eventCount = existing.eventCount + 1)
        } else {
            dict[name] = GedcomPlace(
                id = Uuid.random().toString(),
                name = name,
                normalized = name,
                eventCount = 1
            )
        }
    }

    fun extractYear(dateString: String): Int? {
        val regex = Regex("(\\d{4})")
        val match = regex.find(dateString) ?: return null
        return match.groupValues[1].toIntOrNull()
    }
}
